src/scales/scale.time.js

```diff
@@ -81,6 +81,102 @@ function arrayUnique(items) {
 	return [...set];
 }
 
+function parse(scale, input) {
+	if (helpers.isNullOrUndef(input)) {
+		return null;
+	}
+
+	const adapter = scale._adapter;
+	const options = scale.options.time;
+	const parser = options.parser;
+	let value = input;
+
+	if (typeof parser === 'function') {
+		value = parser(value);
+	}
+
+	// Only parse if its not a timestamp already
+	if (!helpers.isFinite(value)) {
+		value = typeof parser === 'string'
+			? adapter.parse(value, parser)
+			: adapter.parse(value);
+	}
+
+	if (value === null) {
+		return value;
+	}
+
+	if (options.round) {
+		value = scale._adapter.startOf(value, options.round);
+	}
+
+	return +value;
+}
+
+function getDataTimestamps(scale) {
+	const isSeries = scale.options.distribution === 'series';
+	let timestamps = scale._cache.data || [];
+	let i, ilen, metas;
+
+	if (timestamps.length) {
+		return timestamps;
+	}
+
+	metas = scale._getMatchingVisibleMetas();
+
+	if (isSeries && metas.length) {
+		return metas[0].controller._getAllParsedValues(scale);
+	}
+
+	for (i = 0, ilen = metas.length; i < ilen; ++i) {
+		timestamps = timestamps.concat(metas[i].controller._getAllParsedValues(scale));
+	}
+
+	// We can not assume data is in order or unique - not even for single dataset
+	// It seems to be somewhat faster to do sorting first
+	return (scale._cache.data = arrayUnique(timestamps.sort(sorter)));
+}
+
+function getLabelTimestamps(scale) {
+	const isSeries = scale.options.distribution === 'series';
+	const timestamps = scale._cache.labels || [];
+	let i, ilen, labels;
+
+	if (timestamps.length) {
+		return timestamps;
+	}
+
+	labels = scale._getLabels();
+	for (i = 0, ilen = labels.length; i < ilen; ++i) {
+		timestamps.push(parse(scale, labels[i]));
+	}
+
+	// We could assume labels are in order and unique - but let's not
+	return (scale._cache.labels = isSeries ? timestamps : arrayUnique(timestamps.sort(sorter)));
```

Comment:
If using a `Set` for ticks, I don't think we need to do `arrayUnique` anymore here.

Reply:
Hmm. These are the timestamps provided by the user, which are different than the ticks though. The ticks are calculated using these. If we have duplicates here they'll show up in the lookup table, etc. I think I'd rather not change it as it seems risky.